blackjax/inference/hmc/trajectory.py

```diff
@@ -205,8 +205,20 @@ def add_one_state(loop_state):
             new_state = integrator(trajectory.rightmost_state, direction * step_size)
             new_proposal, is_diverging = generate_proposal(initial_energy, new_state)
 
-            new_trajectory = append_to_trajectory(trajectory, new_state)
-            sampled_proposal = sample_proposal(proposal_key, proposal, new_proposal)
+            # At step 0, we always accept the proposal, since we
```

Comment:
I remember you added the extra step to correct a bug in my implementation, but I don't see any of that in the `numpyro` implementation. Did I miss something?

Reply:
I believe here is where numpyro apply the same logic https://github.com/pyro-ppl/numpyro/blob/26905216c0d24b35eee2581fb2a5648730e311e8/numpyro/infer/hmc_util.py#L1018-L1030

Explaination:
```
# After the first doubling we have
o -- o

# Suppose we extend on the right side, the end result would be
         new tree
            /\ 
           /  \
o -- o -- o -- o

# To build the new tree, we take the rightmost from previous tree
o -- x
     ^ select this leaf state

# before this change, we do 1 leapfrog outside of the while_loop
o -- x --> o
           ^ enter the while_loop for tree building, append new leaf after

# now
o -- x
     ^ enter the while_loop for tree building, but at step 0 (the first step) 
       we generate the new tree (ie leave out x) with its associated proposal
```

Reply:
Got it. We can soon write a book on sampling algorithms ^^